scRNA-seq/01-normalizing_scRNA-seq.Rmd

```diff
@@ -92,72 +87,133 @@ plot.
 
-As we should expect, our single-cell RNA-seq transcript counts are largely from 
-a few genes that are expressed a lot, and a lot of genes that are barely 
-expressed at all.
-One way to normalize this data and make it more useful to us, is to log
-transform it.
-These are the default settings for normalization and should work in most 
-instances. 
+In whatever data we are working with, we are always looking to maximize biological 
+variance and minimize technical variance. 
+This is where normalization methods usually come into the workflow.
+For our data, many of these zeroes are likely biologically sensible: an individual
+cell doesn't need to express every gene.
+However, some of these zeroes might be technical in origin: we start with such 
+a small amount of RNA in single cell (we are potentially trying to detect just 
+a few RNA molecules that might exist for this gene in this cell).  
+  
+So ideally, we would like to avoid the technical zeroes, but leave the biological 
+zeroes alone.
+[Lun et al 2016]((https://genomebiology.biomedcentral.com/articles/10.1186/s13059-016-0947-7)) 
+developed a normalization method in `scran` that attempts to do this. 
+  
+*Overall strategy for `scran::computeSumFactors` based normalization:*
+1) Group cells with other like cells by clustering.  
+2) Compute sum factors for each cell cluster grouping.  
+3) Estimate new counts using linear algebra.  
+4) Normalize by the total reads with counts per million reads (CPMs).  
+  
+Because of differences in sequencing depth, it's pretty common to try to correct
+for this by using total per million reads.
+You have probably seen data reported that was, in part, normalized based the 
+total per million reads for a sample such as counts per million reads (CPMs),
+(or TPMs, FPKMs, and RPKMs, which you can [read about more here](https://haroldpimentel.wordpress.com/2014/05/08/what-the-fpkm-a-review-rna-seq-expression-units/) on your own).
+
+#### Normalization step 1: Estimate cell clusters:
+
+```{r}
+# Calculate clusters (can adjust min.size depending on clusters sizes you may 
+# expect based on prior knowledge)
+qclust <- scran::quickCluster(sce_raw, min.size = 100)
+
+# Look at the group sizes before we continue:
+summary(qclust)
+```
+
+#### Normalization step 2: Compute sum factors:
+
+Computer factors for each sample using the clusters we just made.
+
+```{r}
+# Calculate the size factors based on the clusters
+sce_raw <- scran::computeSumFactors(sce_raw, clusters = qclust)
+```
+
+#### Normalization step 3 and 4: Calculate CPMs and normalize using sum factors:
+
+For more details on the normalization [read here](https://rdrr.io/bioc/scran/man/computeSumFactors.html)
 
 ```{r Normalize the Data}
-seurat_norm <- NormalizeData(object = seurat_obj, 
-                             normalization.method = "LogNormalize",
-                             scale.factor = 10000)
+# Normalize the data
+sce_norm <- scater::calculateCPM(sce_raw, use_size_factors = TRUE)
```

Comment:
We had this conversation in person but to document it here -- I was expecting this to be `scater::normalize`. It is possible that these functions are equivalent when using `use_size_factors = TRUE` but let's check. `scater::normalize` also has an option for returning things on a log2 scale which may be helpful.